packages/runtime/container-runtime/src/dataStoreContext.ts

```diff
@@ -373,6 +373,13 @@ export abstract class FluidDataStoreContext extends TypedEventEmitter<IFluidData
         const channel = await factory.instantiateDataStore(this, existing);
         assert(channel !== undefined, 0x140 /* "undefined channel on datastore context" */);
         this.bindRuntime(channel);
+
+        if (!existing) {
```

Comment:
why is this needed?
I remember it used to be that we had to do similar operations to ensure DataObject has a change to initialize before we send attach op with its content. But currently attach of data store happens when its handle gets references somewhere in the graph, and in order to get a handle someone will need to realize it. Plus, DataObjectFactory has explicit flow to ensure initialization is done before use.

Is the issue here - aliasing before DataObject was fully initialized?

Reply:
@vladsud I'll ping you to discuss in a call, but wanted to drop some thoughts here for now:

For `DataObjectFactory`/`DataObject` it does seem like we could leverage `finishInitialization()` and/or `initializeInternal()` to ensure the entrypoint is fully built, but other factories we already have (`TestFluidObjectFactory`, `LazyLoadedDataObjectFactory`, `AgentSchedulerFactory`, or the ones in some of our examples: `SmdeFactory`, `ProseMirrorFactory`) don't have the same flow. And factories built by consumers wouldn't necessarily have it either. So for those cases it seems like we'd want to have the framework guarantee that the entrypoint-initialization function added in this PR will get called automatically, and particularly that it'll be given a chance to run before the data store is attached (for the same [reason as in PureDataObjectFactory](https://github.com/microsoft/FluidFramework/blob/35af160af92fce0ae3e41f13ac376db4ca6127fc/packages/framework/aqueduct/src/data-object-factories/pureDataObjectFactory.ts#L84-L92)).

An alternative could be to make the (non-breaking) API change in this PR expect a fully-built entrypoint object instead of an async function to build it, but that defeats the purpose of the initialization flow for DataObjectFactory, right? Which seems to want to delay that construction as much as possible.

That said, I have a [draft PR](https://github.com/microsoft/FluidFramework/pull/12239) where I've experimented with commenting the changes in this file. The one in `realizeCore` (around line 380) don't seem to be necessary for our tests to pass. However if I comment the one in `attachRuntime()` (around line 1051), [this assert](https://github.com/microsoft/FluidFramework/blob/a5d074748aafee897cdda390ffe8f6b9b83a1913/packages/test/local-server-tests/src/test/noDeltaStream.spec.ts#L97) fails (plus other errors I haven't investigated in detail but are probably a similar issue). Without the two new lines, we somehow end in a state where, by the time an op gets sent, the data store is already attached, that causes the connection to go from readonly to write (passing through a disconnect first), the assert consistently catches the data store runtime in that transient disconnected state, and fails. I'm still not entirely clear on why we end up attached without the new lines but not with them.

Reply:
Maybe it's fine, but I do not like conditional code like that because it's very hard to assess the impact on various code flows. For example - if what you care about is important on creation flow, is it important (or not) in summarization flow, where realize() would happen, but handle is not going to be used? Is anything that runs as result of resolving this handle important to the state of the object and its lifecycle? How do we answer these questions?

It's not about calling this code, it's more about it making conditional on some existing flag. Note that exising === false when data store is loaded in case of saved detached container state. So, there is a path where attach will follow after running this code with existing == false. Maybe it's Ok in most cases because prior to that (in some previous session) we already went through this code with existing === false, but I have hard time evaluating this matrix.

So I'd rather see code like this unconditional, but comment mention that we definitely need it for creation flow, and making it uniform for all others.

That all said, I have hard time understanding connection to assert you mentioned. On the surface it looks unrelated, and it might be worth drilling into it to understand better - there might be unrelated bugs hiding somewhere.

Reply:
One thing to point out (realized based on this comment thread, as well as [this comment](https://github.com/microsoft/FluidFramework/pull/11887/files#diff-c40fee81fe1d30fba831dd5cf1189760d40e78b67cc7b8d42a25740a77271631L84)

On one hand, it sounds like we can remove such code in DataStore, as you are shifting responsibility to container runtime layer to get entry point (and thus fully initialize objects). On another hand, code in DataObject layer is very explicit (with comment above) and relying on container runtime behavior feels dangerous.

But more important - This is more a directional comment, as we are not there yet (and no clue when we will get there). But factories should be used for loading, not for creation. Creation should mimic what DataObject does - be direct (using higher-level factory, not data store factory), and fully initialized result (data store runtime) should be attached to context. All the examples that you mention, ideally, should be changed to follow this pattern and over time we should push our customers also to not conflict load & create pattern.

I do not know how hard it is to start and where to draw compromise, but essentially any code that breaks when this code in context is removed should be changed to leverage detached creation flow.